class-two-factor-core.php

```diff
@@ -1433,11 +1612,63 @@ public static function user_two_factor_options( $user ) {
 			$primary_provider_key = null;
 		}
 
-		wp_nonce_field( 'user_two_factor_options', '_nonce_user_two_factor_options', false );
+		// DEBUG START - This section is not intended to be in the final release.
+		if ( $is_current_user ) {
+			$session = WP_Session_Tokens::get_instance( $user->ID )->get( wp_get_session_token() );
+
+			if ( self::is_current_user_session_two_factor() ) {
+				printf(
+					'<div class="notice notice-info inline"><p>%s</p></div>',
+					sprintf(
+						'You are currently logged in with a session from %s. <br>' .
+							'Your 2FA token was last confirmed at %s.<br>' .
+							'Your 2FA method for login was %s (%s after login, %s ago).',
+						'<code>' . date( 'r', $session['login'] ) . '</code>',
+						'<code>' . date( 'r', $session['two-factor-login'] ) . '</code>',
+						'<code>' . esc_html( $session['two-factor-provider'] ) . '</code>',
+						human_time_diff( $session['login'], $session['two-factor-login'] ),
+						human_time_diff( time(), $session['two-factor-login'] ),
+					)
+				);
+			} else {
+				printf(
+					'<div class="notice notice-warning inline"><p>%s</p></div>',
+					sprintf(
+						'You are currently logged in with a session from %s. <br>' .
+							'<strong>You are NOT using a Two Factor session.</strong>',
+						'<code>' . date( 'r', $session['login'] ) . '</code>'
+					)
+				);
+			}
+		}
+		// DEBUG END
+
+		// This is specific to the current session, not the displayed user.
+		$show_2fa_options = self::current_user_can_update_two_factor_options();
 
+		if ( ! $show_2fa_options ) {
+			$url = self::get_user_two_factor_revalidate_url();
+			$url = add_query_arg( 'redirect_to', urlencode( self::get_user_settings_page_url( $user->ID ) . '#two-factor-options' ), $url );
+
+			printf(
+				'<div class="notice notice-warning inline"><p>%s</p></div>',
+				sprintf(
+					__( 'To update your Two-Factor options, you must first revalidate your session.', 'two-factor' ) .
+					'<br><a class="button" href="%s">' . __( 'Revalidate now', 'two-factor' ) . '</a>',
+					esc_url( $url )
+				)
+			);
+		}
+
+		printf(
+			'<fieldset id="two-factor-options" %s>',
+			$show_2fa_options ? '' : 'disabled="disabled"',
+		);
```

Comment:
I'm not too fond of this UI, but without recreating the entire UI this seems to be the simplest way forward.

I had planned on using the "interim login" feature to display in a thickbox-style overlay, but the core re-authentiacte-your-session implementation isn't extendable and would need to be duplicated into the plugin. Likewise, the interim-login code only has support for the Customizer, not for other JS clients.